### PR TITLE
Added whylogs integration and connected to metadata preprocessing

### PR DESCRIPTION
**Pull request by anmshkmr**
# Code Pull Requests

Please provide the following:

- a clear explanation of what your code does
- if applicable, a reference to an issue
- a reproducible test for your PR (code, config and data sample)

# Documentation Pull Requests

Note that the documentation HTML files are in `docs/` while the Markdown sources are in `mkdocs/docs`.

If you are proposing a modification to the documentation you should change only the Markdown files.

`api.md` is automatically generated from the docstrings in the code, so if you want to change something in that file, first modify `ludwig/api.py` docstring, then run `mkdocs/code_docs_autogen.py`, which will create `mkdocs/docs/api.md` .

### CODE REVIEW

File: ludwig/data/preprocessing.py [hunk 1]
```diff
@@ -1383,7 +1454,8 @@ def preprocess_for_training(
         skip_save_processed_input=False,
         preprocessing_params=default_preprocessing_parameters,
         backend=LOCAL_BACKEND,
-        random_seed=default_random_seed
+        random_seed=default_random_seed,
+        callbacks=[]
```
Comment (tgaddair) [inline ludwig/data/preprocessing.py:1458]: In general, we should avoid using `[]` as a default argument in Python, as it becomes a mutable constant, which can have unexpected side effects when you call the function multiple times. Better alternative is to use `None`.

File: ludwig/data/preprocessing.py [hunk 2]
```diff
@@ -1089,6 +1157,9 @@ def build_dataset(
         backend
     )
 
+    for callback in callbacks or []:
+        callback.on_build_metadata_end(dataset_df)
+
```
Comment (tgaddair) [inline ludwig/data/preprocessing.py:1162]: Can add callbacks for `on_build_data` below.

File: ludwig/contribs/whylogs/__init__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+class WhyLogsCallback(Callback):
+    def __init__(self, path_to_config=None):
+        self.session = get_or_create_session(path_to_config)
+        self.logger = self.session.logger()
+
+    def on_build_metadata_start(self, df):
+        self.logger.log({'metadata_start': df})
+
+    def on_build_metadata_end(self, df):
+        self.logger.log({'metadata_end': df})
```
Comment (tgaddair) [inline ludwig/contribs/whylogs/__init__.py:14]: We probably only need to log once during `on_build_metadata_start`. The other callbacks we can just leave for others who may be interested in overriding them.

**Review by tgaddair — COMMENTED** [review]
Nice! As discussed, tests and distinguishing between train and predict modes are good next steps, then we should be good to land.

File: ludwig/contribs/whylogs/__init__.py (hunk 3 above)
Comment (anmshkmr) [reply ludwig/contribs/whylogs/__init__.py]: Makes sense. This in combo with change from your previous comment makes sense.

File: ludwig/data/preprocessing.py (hunk 2 above)
Comment (anmshkmr) [reply ludwig/data/preprocessing.py:1182]: Done.

**Comment by anmshkmr** [comment]
Took out the WIP tag. Tested locally and verified we're seeing whylog proto bins in the directory where we run tests.

File: requirements.txt [hunk 4]
```diff
@@ -22,6 +22,7 @@ tensorboard
 torchmetrics>=0.6.0
 torchinfo
 filelock
+whylogs
```
Comment (tgaddair) [inline requirements.txt:25]: We should remove this from the main requirements, as it is optional.

File: ludwig/data/preprocessing.py [hunk 5]
```diff
@@ -981,7 +998,8 @@ def preprocess_for_training(
             skip_save_processed_input=False,
             preprocessing_params=default_preprocessing_parameters,
             backend=LOCAL_BACKEND,
-            random_seed=default_random_seed
+            random_seed=default_random_seed,
+            callbacks=[]
```
Comment (tgaddair) [inline ludwig/data/preprocessing.py:1002]: Let's replace `[]` with `None` to avoid mutable default issue.

File: ludwig/data/preprocessing.py [hunk 6]
```diff
@@ -1113,7 +1132,8 @@ def build_dataset(
         backend=LOCAL_BACKEND,
         random_seed=default_random_seed,
         skip_save_processed_input=False,
-        callbacks=[]
+        callbacks=[],
+        context=None
```
Comment (tgaddair) [inline ludwig/data/preprocessing.py:1136]: `context` can get overloaded, so maybe something like `mode` will be more natural.

File: ludwig/callbacks.py [hunk 7]
```diff
@@ -116,16 +116,16 @@ def on_test_start(self, trainer, progress_tracker, save_path):
     def on_test_end(self, trainer, progress_tracker, save_path):
         pass
 
-    def on_build_metadata_start(self, df):
+    def on_build_metadata_start(self, df, dataset_name):
```
Comment (tgaddair) [inline ludwig/callbacks.py:119]: `dataset_name` should be consistent with the other param if possible, so maybe we can change this to `mode`.

**Review by tgaddair — COMMENTED** [review]
Looks great! Last thing we need to nail down is how this works with Dask. We can sync up for a quick meeting to discuss.

File: ludwig/data/preprocessing.py (hunk 6 above)
Comment (anmshkmr) [reply ludwig/data/preprocessing.py:1136]: Done.

File: ludwig/data/preprocessing.py (hunk 5 above)
Comment (anmshkmr) [reply ludwig/data/preprocessing.py:1002]: Done.

File: ludwig/callbacks.py (hunk 7 above)
Comment (anmshkmr) [reply ludwig/callbacks.py:119]: Done.

**Review by tgaddair — APPROVED** [review]
LGTM!

**Pull request merged**